### PR TITLE
Feature/rhino object functions

### PR DESCRIPTION
**Pull request by robin-oval**
Rhino curve and surface functions
curve length and is_closed
surface kinks, closest_point, closest_point_on_boundary and several mapping/remapping functions between XYZ and UV(0) spaces based on the surface's parameterisation

### CODE REVIEW

File: src/compas_rhino/geometry/curve.py [hunk 1]
```diff
@@ -65,6 +65,17 @@ def from_selection(cls):
...
+    def length(self):
+        """Return the length of the curve.
+
+        Returns
+        -------
+        float
+            The curve's length.
+
+        """
+        return rs.CurveLength(self.guid)
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/curve.py:77]: the object is also stored, and there is a method of the object for getting the length...
https://developer.rhino3d.com/api/RhinoCommon/html/M_Rhino_Geometry_Curve_GetLength_2.htm

File: src/compas_rhino/geometry/curve.py [hunk 2]
```diff
@@ -101,6 +112,18 @@ def is_polyline(self):
...
+        """Assess if the curve is closed.
+
+        Returns
+        -------
+        bool
+            True if the curve is closed. False otherwise.
+
+        """
+
+        return rs.IsCurveClosed(self.guid)
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/curve.py:125]: same comment
https://developer.rhino3d.com/api/RhinoCommon/html/P_Rhino_Geometry_Curve_IsClosed.htm

File: src/compas_rhino/geometry/surface.py [hunk 3]
```diff
@@ -229,6 +234,33 @@ def borders(self, type=1):
...
+    def kinks(self, threshold=1e-3):
+        """Return the XYZ coordinates of kinks, i.e. tangency discontinuities, along the surface's boundaries.
+
+        Returns
+        -------
+        list
+            The list of XYZ coordinates of surface boundary kinks.
+
+        """
+
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: remove empty line

File: src/compas_rhino/geometry/surface.py [hunk 4]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+            XYZ coordinates.
+
+        Returns
+        -------
+        list
+            The XYZ coordinates of the closest point on the boundaries of the surface.
+
+        """
+
+        borders = self.borders(type = 0)
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: spaces

File: src/compas_rhino/geometry/surface.py [hunk 5]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+        proj_dist = {tuple(proj_xyz): distance_point_point(xyz, proj_xyz) for proj_xyz in [RhinoCurve(border).closest_point(xyz) for border in borders]}
+        delete_objects(borders)
+        return min(proj_dist, key = proj_dist.get)
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: spaces

File: src/compas_rhino/geometry/surface.py [hunk 6]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+            XYZ coordinates.
+
+        Returns
+        -------
+        list
+            The UV(0) coordinates of the mapped point.
+
+        """
+
+        return rs.SurfaceClosestPoint(self.guid, xyz) + (0.,)
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py:365]: not sure i understand what this function does

File: src/compas_rhino/geometry/surface.py [hunk 7]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+        uv : list
+            UV(0) coordinates.
+
+        Returns
+        -------
+        list
+            The XYZ coordinates of the re-mapped point.
+
+        """
+
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: blank line

File: src/compas_rhino/geometry/surface.py (hunk 6 above)
Comment (tomvanmele) [reply src/compas_rhino/geometry/surface.py:365]: blank line

File: src/compas_rhino/geometry/surface.py [hunk 8]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+        xyz : list
+            XYZ coordinates.
+
+        Returns
+        -------
+        list
+            The XYZ coordinates of the closest point on the boundaries of the surface.
+
+        """
+
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: blank line

File: src/compas_rhino/geometry/surface.py [hunk 9]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+        xyz : list
+            XYZ coordinates.
+
+        Returns
+        -------
+        list
+            The XYZ coordinates of the closest point on the surface.
+
+        """
+
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py:319]: blank line

File: src/compas_rhino/geometry/surface.py [hunk 10]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+        Returns
+        -------
+        list
+            The UV(0) coordinates of the mapped point.
+
+        """
+
+        return rs.SurfaceClosestPoint(self.guid, xyz) + (0.,)
+
+    def remap_xyz_point(self, uv):
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: example?

File: src/compas_rhino/geometry/surface.py [hunk 11]
```diff
@@ -278,7 +303,143 @@ def pull_mesh(self, mesh, fixed=None, d=1.0):
...
+
+        return tuple(rs.EvaluateSurface(self.guid, *uv))
+
+    def remap_xyz_line(self, line):
+        """Return the XYZ points from the re-mapping of a UV line based on the UV parameterisation of the surface.
+
+        Parameters
+        ----------
+        uv : list
+            List of UV(0) coordinates.
```
Comment (tomvanmele) [inline src/compas_rhino/geometry/surface.py]: UV(0)

**Review by tomvanmele — CHANGES_REQUESTED** [review]
we should avoid just providing wrappers to the corresponding rs.xxx functions and use RhinoCommon object methods where possible...

it is not entirely clear what some of the functions actually do, so examples and more explanations are necessary...

**Comment by robin-oval** [comment]
instead of wrapping rhinoscriptsyntax functions, is it better to wrap RhinoCommon functions or not wrapping anything and calling RhinoCommon methods directly?

**Review by tomvanmele — APPROVED** [review]

**Pull request merged**